### PR TITLE
Improve macro-op fusion: skip nop and add lui + addi

### LINKED ISSUE #177: `do_nop` is not negligible
`do_nop` appears to consume some CPU cycles according to perf report. It is generated by macro operation fusion, and we should aim to eliminate its overhead as early as possible.

Reproduce:
1. Apply the following patch:
```diff
--- a/Makefile
+++ b/Makefile
@@ -5,6 +5,7 @@ OUT ?= build
 BIN := $(OUT)/rv32emu
 
 CFLAGS = -std=gnu99 -O2 -Wall -Wextra
+CFLAGS += -g -fno-omit-frame-pointer
 CFLAGS += -Wno-unused-label
 CFLAGS += -include src/common.h
 
@@ -80,7 +81,7 @@ endif
 
 # For tail-call elimination, we need a specific set of build flags applied.
 # FIXME: On macOS + Apple Silicon, -fno-stack-protector might have a negative impact.
-$(OUT)/emulate.o: CFLAGS += -fomit-frame-pointer -fno-stack-check -fno-stack-protector
+$(OUT)/emulate.o: CFLAGS += -fno-stack-check -fno-stack-protector
 
 # Clear the .DEFAULT_GOAL special variable, so that the following turns
 # to the first target after .DEFAULT_GOAL is not set.
```
2. Rebuild and run dhrystone benchmark
```shell
$ make clean all
$ perf record -g build/rv32emu build/dhrystone.elf
```
3. Check the report via `perf report -g` and be aware of the percentage of `do_nop`.
```
-   99.87%     0.16%  rv32emu  rv32emu             [.] main
   - 99.71% main
      + 68.62% rv_step
        8.11% do_addi
        3.05% do_nop
        3.05% do_auipc.part.0
        2.26% block_find
        1.96% do_bne
        1.65% do_sw
        1.60% do_lw
        1.24% do_bgeu
        1.18% do_beq
        1.05% do_or
        0.95% do_fuse3
        0.83% do_andi
        0.82% do_lbu
        0.72% do_jal
```

### PR DESCRIPTION
**Pull request by qwe661234**
1. Refine origin fused instruction by skipping insturction nop and correctly updating value to register.
2. Add new fused insturction lui + addi.

Benchmark dhrystone gains about 3% performance improvement base on this modification.

Close: #177

### CODE REVIEW

File: src/emulate.c [hunk 1]
```diff
@@ -312,15 +312,7 @@ static uint32_t last_pc = 0;
...
-/* Internal */
-static bool do_nop(riscv_t *rv, const rv_insn_t *ir)
-{
-    rv->X[rv_reg_zero] = 0;
-    rv->csr_cycle++;
-    rv->PC += ir->insn_len;
-    const rv_insn_t *next = ir + 1;
-    MUST_TAIL return next->impl(rv, next);
-}
+RVOP(nop, {/* no operation */})
```
Comment (jserv) [inline src/emulate.c]: Why did you change `nop`?

**Comment by jserv** [comment]
Are you ready to close #177 ?

**Comment by qwe661234** [comment]
> Are you ready to close #177 ?

Yes, we cannot find `do_nop` in perf report after this modification.
